static/app/components/hovercard.tsx

```diff
@@ -65,82 +73,44 @@ type Props = DefaultProps & {
-      portal.setAttribute('id', 'hovercard-portal');
-      document.body.appendChild(portal);
-    }
-    this.portalEl = portal;
-    this.tooltipId = domId('hovercard-');
-    this.scheduleUpdate = null;
-  }
-
-  state: State = {
-    visible: false,
-  };
-
-  componentDidUpdate(prevProps: Props) {
-    const {body, header} = this.props;
-
-    if (body !== prevProps.body || header !== prevProps.header) {
-      // We had a problem with popper not recalculating position when body/header changed while hovercard still opened.
-      // This can happen for example when showing a loading spinner in a hovercard and then changing it to the actual content once fetch finishes.
-      this.scheduleUpdate?.();
-    }
-  }
+  children: React.ReactNode;
+}
 
-  portalEl: HTMLElement;
-  tooltipId: string;
-  hoverWait: number | null = null;
-  scheduleUpdate: (() => void) | null;
+function Hovercard(props: HovercardProps): React.ReactElement {
+  const [visible, setVisible] = React.useState(false);
 
-  handleToggleOn = () => this.toggleHovercard(true);
-  handleToggleOff = () => this.toggleHovercard(false);
+  const inTimeout = React.useRef<number | null>(null);
+  const scheduleUpdateRef = React.useRef<(() => void) | null>(null);
 
-  toggleHovercard = (visible: boolean) => {
-    const {displayTimeout} = this.props;
+  const portalEl = React.useMemo(() => findOrCreatePortal(), []);
+  const tooltipId = React.useMemo(() => domId('hovercard-'), []);
 
-    if (this.hoverWait) {
-      clearTimeout(this.hoverWait);
+  React.useEffect(() => {
+    // We had a problem with popper not recalculating position when body/header changed while hovercard still opened.
+    // This can happen for example when showing a loading spinner in a hovercard and then changing it to the actual content once fetch finishes.
+    if (scheduleUpdateRef.current) {
+      scheduleUpdateRef.current();
     }
-
-    this.hoverWait = window.setTimeout(() => this.setState({visible}), displayTimeout);
-  };
-
-  render() {
-    const {
-      bodyClassName,
-      containerClassName,
-      className,
-      header,
-      body,
-      position,
-      show,
-      tipColor,
-      tipBorderColor,
-      offset,
-      modifiers,
-    } = this.props;
-
-    // Maintain the hovercard class name for BC with less styles
-    const cx = classNames('hovercard', className);
-    const popperModifiers: PopperProps['modifiers'] = {
+  }, [props.body, props.header]);
+
+  const toggleHovercard = React.useCallback(
+    (value: boolean) => {
+      // If a previous timeout is set, then clear it
+      if (typeof inTimeout.current === 'number') {
+        clearTimeout(inTimeout.current);
+      }
+
+      // Else enqueue a new timeout
+      inTimeout.current = window.setTimeout(
+        () => setVisible(value),
+        props.displayTimeout ?? 100
+      );
+    },
+    [props.displayTimeout]
+  );
+
+  const popperModifiers = React.useMemo(() => {
```

Comment:
do we do an expensive calculation here that requires `useMemo`?

Reply:
No, but we are creating a new object reference each time and it would cause popper to rerender (or do a deep props comparison in case it wants to avoid it)

Reply:
I think we should do some education session, create a document, or something around this, because as far as I know we try to avoid using them, but maybe we should use them more to avoid problems like the one you mentioned.

Reply:
Oh, I had no idea 🤔 I think that would go against best react practices tbh. To be fair, it is currently very easy to make mistakes in our codebase because our linters do not enforce exhaustive hooks (see see exhaustive-deps rule https://reactjs.org/docs/hooks-rules.html#eslint-plugin), which makes it easy to forget to add proper dependencies and you end up with stale references...

Reply:
I see... I'm not sure why we don't have this rule enabled yet...but if this is beneficial for us...let's add it ^^

https://github.com/getsentry/eslint-config-sentry